### PR TITLE
Text editor: rotate

### PR DESCRIPTION
**Pull request by hius07**
Closes https://github.com/koreader/koreader/issues/12622.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12658)

### CODE REVIEW

**Comment by hius07** [comment]
It should be tested on a gyro device. @mergen3107 can you help?

**Comment by mergen3107** [comment]
When I go back to Text Editor, it is again in un-rotated state (`D`).

**Comment by mergen3107** [comment]
Is this what you wanted to test, or anything else?

**Comment by mergen3107** [comment]
While in Text Editor, I rotated Kindle physically by 90 deg.
Text Editor did not rotate.
Exited from it, and FM (with top menu still open where I opened a file) is still open and did not rotate neither.

**Comment by mergen3107** [comment]
I think, it would be much more handy to have Text Editor rotate by gyro, the same way as History. This (I think) should apply to all windows that can rotate. This will make it consistent in terms of UX

**Comment by hius07** [comment]
Thanks, manual rotation works as expected.
I've added gyro support, can you try?

**Comment by hius07** [comment]
One more try please.
[main.lua.txt](https://github.com/user-attachments/files/17451159/main.lua.txt)

**Comment by mergen3107** [comment]
Nope, doesn't work neither.

I tried doing `logger.dbg` inside this function, and it doesn't show up at all.
So this event is not firing up

**Comment by mergen3107** [comment]
Is `rotation` passed to the Text Editor instance correctly?

**Comment by mergen3107** [comment]
Sorry, I take this back:
```
10/20/24-11:58:45 DEBUG TEXTEDITOR: rotation =  2 
```
2 is my current rotation (D - upside down).

However, it doesn't register when I rotate it to other ones while in Text Editor

**Comment by mergen3107** [comment]
Sorry again :D 
The log above was when I tried previous iteration.

Now I don't see it at all

**Comment by Commodore64user** [comment]
I wish you could add `Menu` key event to it.

**Comment by hius07** [comment]
@mergen3107 thanks for testing, now it should work (tesetd with the httpInspector).

**Comment by hius07** [comment]
Please note that if you have rotated TE manually, after closing TE the file manager remains in non-rotated orientation.
If you want FM to follow the TE rotation, rotate the device.

**Comment by mergen3107** [comment]
@hius07 
It did work really well:
1) Gyro rotation in TE worked
2) However, there is top menu draw over TE, so part of the TE text is not visible.
3) Button in the top left corner works.
4) If I use to rotate, the rotation window is still drawn, but doesn't work in TE (until I tap in the  corner again).
5) However, once I closed TE, I saw book page (where I was when launched TE), then Top Menu, then little window with rotations. That little window was also dismissible when I tapped outside. Also, top menu was also dismissible.

So it seems like layers of what is rendered and what is interactable got mismatched. See screenshots and logs (don't know if they help):
[crash_TE.txt](https://github.com/user-attachments/files/17465003/crash_TE.txt)
Reader_2024-10-21_133943
Reader_2024-10-21_133949

**Comment by hius07** [comment]
Have you applied changes to all 4 files? I cannot reproduce the top menu over the Text editor.

**Comment by mergen3107** [comment]
Yes. After each patch I revert it.

This time it said it modified four different files OK.

**Comment by mergen3107** [comment]
Top Menu is there because I go to Top Menu - Text Editor - and tap on a file

**Comment by hius07** [comment]
Okay, I think that you press the rotation button in TE, it rotates, then you rotate the device and the gyro forces the rotation of all the underlying content (reader and the top menu) that is drawn over TE.
What if you manually rotate TE and **do not rotate** the device?

**Comment by mergen3107** [comment]
> What if you manually rotate TE and do not rotate the device? Do you see the top menu?

If I do just this, then no - all good. Upon exit though, reader is still in old orientation (is this what you wanted?)

**Comment by hius07** [comment]
Thanks, yes. Also I can reproduce the top menu overdrawn. Diggin into it.

**Comment by hius07** [comment]
The description https://github.com/koreader/koreader/commit/2c33fc65769c59d2a3f95670df81d67e6e657c70 is true, upper menues don't support rotations, so reverting the rough changes from https://github.com/koreader/koreader/pull/11297.

**Comment by hius07** [comment]
@mergen3107 now it works well, manually and automatically, syncing them.
But for gyro devices manual rotation doesn't have much sense.

**Comment by mergen3107** [comment]
@hius07 
Tested, works great!

It also keeps landscape rotation when I exit from TE.
However, it doesn't honor `Keep current rotation across views` being off.

Also, I got a crash when exiting from rotated PDF to FM:
```
./luajit: frontend/apps/filemanager/filemanager.lua:738: attempt to index field 'file_chooser' (a nil value)
stack traceback:
	frontend/apps/filemanager/filemanager.lua:738: in function 'reinit'
	frontend/apps/filemanager/filemanager.lua:69: in function 'onSetRotationMode'
	frontend/apps/filemanager/filemanager.lua:87: in function 'setRotationMode'
	frontend/apps/filemanager/filemanager.lua:1227: in function 'showFiles'
	frontend/apps/reader/readerui.lua:581: in function 'showFileManager'
	frontend/apps/reader/modules/readermenu.lua:50: in function 'callback'
	frontend/ui/widget/touchmenu.lua:784: in function 'switchMenuTab'
	frontend/ui/widget/touchmenu.lua:405: in function 'callback'
	frontend/ui/widget/iconbutton.lua:137: in function 'handleEvent'
	frontend/ui/widget/container/inputcontainer.lua:266: in function 'handleEvent'
	frontend/ui/widget/container/widgetcontainer.lua:83: in function 'propagateEvent'
	...
	frontend/ui/widget/container/widgetcontainer.lua:101: in function 'handleEvent'
	frontend/ui/widget/container/widgetcontainer.lua:83: in function 'propagateEvent'
	frontend/ui/widget/container/widgetcontainer.lua:101: in function 'handleEvent'
	frontend/ui/uimanager.lua:911: in function 'sendEvent'
	frontend/ui/uimanager.lua:53: in function '__default__'
	frontend/ui/uimanager.lua:1438: in function 'handleInputEvent'
	frontend/ui/uimanager.lua:1538: in function 'handleInput'
	frontend/ui/uimanager.lua:1582: in function 'run'
	./reader.lua:265: in main chunk
	[C]: at 0x00013ead
lipc-wait-event exited normally with status: 0
```
Is this related?

**Comment by mergen3107** [comment]
> But for gyro devices manual rotation doesn't have much sense.

True, but it doesn't hurt being in the top left corner.
Maybe someone doesn't use gyro at all, so that's a good alternative to going back to bottom menu to rotate and back to TE

**Comment by mergen3107** [comment]
Yes, I think the crash above is related to this patch. Without it, I have to crash.

**Comment by mergen3107** [comment]
> However, it doesn't honor `Keep current rotation across views` being off.

In that case, for example, when I enter TE from Portrait book, then I rotate TE to Landscape, and then exit back to book - book should stay Portrait. Now it is rotate to Landscape with or without the checkbox.

**Comment by hius07** [comment]
> Keep current rotation across views

This setting is not honored by TE, because "views" are FM or Reader.

Thanks, I'll look into the crash.

**Comment by mergen3107** [comment]
Ah, I see. So they (R/TE, or FM/TE) will share it.

**Comment by hius07** [comment]
Summary of changes in the core: since the upper widget now can rotate independently of the base (FM or Reader), I introduce and maintain `self.rotation_mode` in the base, that can differ from `Screen:getRotationMode()`.

**Comment by hius07** [comment]
Because I want to rotate the Text Editor, keeping the underlying FM not rotated.
So I call `Screen:getRotationMode()`, reinit TE, but do not reinit FM.
But then, a gyro device can be rotated itself, and we need to know how to handle the `SetRotationMode` event.

**Comment by NiLuJe** [comment]
> Because I want to manually rotate the Text Editor, keeping the underlying FM not rotated.

New question, then: *why*?

Wouldn't be easier to just `nil` the FM's rotation handler and restore it on exit from TE?

**Comment by hius07** [comment]
Because of gyro devices (as discussed earlier here with @mergen3107): open TE over FM, manually rotate TE, rotate the device -> FM must be rotated (responding to gyro).

**Comment by NiLuJe** [comment]
> manually rotate TE

Make sure a manual rotate rotate the parent widget at the same time?

**Comment by hius07** [comment]
It takes time, especially for Reader.
And on non-gyro devices we must restore the initial orientation after closing TE, more time.

**Comment by hius07** [comment]
I expect we will make more widgets rotatable (by gyro and manually for non-gyro devices).

**Comment by mergen3107** [comment]
@hius07 
All tests successful!
Gyro works, no crashes when exiting to FM.

Manual rotation works too, but doesn't affect Reader/FM (fine by me).

**Comment by NiLuJe** [comment]
Then I would rather limit the dame to TE, and just resync parent widget to TE rota on exit. It's going to be a NOP 99.9% of the time anyway ;).

**Comment by mergen3107** [comment]
> It's going to be a NOP 99.9% of the time anyway ;)

And you'll read my ticket about the rest 0.01% case in 2 min :D

**Comment by hius07** [comment]
I think that my "damage" is less than the proposed one.
(1) On non-gyro devices, the parent widget is not rotated when rotating TE manually. On closing TE we see the parent widget in the initial orientation.
(2) On gyro devices, both TE and the parent widget are rotated in sync by the gyro event. On closing TE we immediately see the parent widget in current device orientation.
(3) On gyro devices, the only special case is: rotate TE manually and **then** rotate the device. Only in this case I use parent's `self.rotation_mode` to sync it with the screen rotation mode, if needed. On closing TE we immediately see the parent widget in current device orientation, same as (2).

**Comment by hius07** [comment]
Anyway, I'll try another way: never rotate the parent widget when rotating TE, and sync the parent widget on closing TE. Will compare the "damage".

**Comment by NiLuJe** [comment]
My main gripe about it is that I really really don't want people to have to thing about supporting another way to access the rotation status, so the less whatever approach ends up being necessary leaks through elsewhere, the better ;).

**Comment by hius07** [comment]
The question **for gyro devices** is: should we rotate the underlying parent widget (FM/Reader) when rotating the upper widget by gyro events?

We rotate it under History and Collections.
I propose to rotate it under TextEditor. You propose to not.
They all are fullscreen widgets, so actually it doesn't matter at the moment.

But if we are going to extend support of gyro devices, we can add rotation of, for example, Go to page dialog or Skim widget.
In this cases we **must** rotate the parent widget.
So we should start now with a consistent decision, taking into account future development.

@mergen3107 what do you think as a user of a gyro device, is it useful to have more dialogs rotatable? We can encounter unexpected difficulties in implementing this (for example, InputDialog over Bookmark list over Reader, or InputDialog over the upper menu), so is it worth to make efforts?

**Comment by Frenzie** [comment]
> We rotate it under History and Collections.
I propose to rotate it under TextEditor. You propose to not.

There's no hard to follow the code path in History and Collections though?

**Comment by hius07** [comment]
FM now
https://github.com/koreader/koreader/blob/42d3f25034778c121dc458b3d5246896fedc5db2/frontend/apps/filemanager/filemanager.lua#L63-L71
FM proposed
https://github.com/koreader/koreader/blob/4d9b2ac6e7509b3a614e755ca5fc7250323329f5/frontend/apps/filemanager/filemanager.lua#L63-L76

**Comment by mergen3107** [comment]
> The question for gyro devices is: should we rotate the underlying parent widget (FM/Reader) when rotating the upper widget by gyro events?

IMHO, if that's a full screen widget, it should just follow the setting `Keep current orientation across views`. If it is checked, then yes, it rotates along. If it unchecked, only widget rotates. Might be useful, say, when editing long lines in TE (so need landscape), while still reading in Portrait.

If that's a small window widget, then I think no. Especially if it is an Input one - I wouldn't like to have keyboard run away if I rotate it by accident :D 

Plus, it will make things complicated because in Reader I have auto profiles? Single/double column rendering takes time, and will take time when rotating small window widgets + Reader as well.

**Comment by NiLuJe** [comment]
Yup, perfect agreement with what @mergen3107 just said ;).

**Comment by NiLuJe** [comment]
Sidebar about [4d9b2ac](https://github.com/koreader/koreader/pull/12658/commits/4d9b2ac6e7509b3a614e755ca5fc7250323329f5)

Unless I missed it, the actual module file still exists, so I'd get rid of it, too ;).

**Comment by hius07** [comment]
> IMHO, if that's a full screen widget, it should just follow the setting `Keep current orientation across views`.

So History/Collections' rotation isn't good currently?
It doesn't follow the setting.

**Comment by NiLuJe** [comment]
That's a good point.

One could argue that anything FM-related should behave like FM itself ;).

That setting exists mostly because Reader *can* rotate on its own, because of document-specific rotations saved in docsettings.

Sooo, yeah, I'd probably amend that to say that anything FM-related should *always* track the current FM rotation.

**Comment by Frenzie** [comment]
> One could argue that anything FM-related should behave like FM itself ;).

I don't think they're FM-related (except in a technical sense).

But from the above piece of code it sounds like they're doing whatever they're doing as a side effect because the menu follows rotation? I doubt that's by design for such full screen menus.

**Comment by mergen3107** [comment]
> So History/Collections' rotation isn't good currently?
It doesn't follow the setting.

I haven't noticed it because I rarely rotate History. It is the books that make my head (and Kindle) spin :D

**Comment by mergen3107** [comment]
I always vouch for consistency in UX, so I would say that History/Collections is a full screen widget and should honor the setting :D 

Because you can open History/Collections both from Reader and FM, so I don't think it is more FM related.

**Comment by NiLuJe** [comment]
> a saved global preference

Which reminded me that `fm_rotation_mode` is a thing that exists, and I really wouldn't want stuff like that to proliferate, so I'd rather stuff that's not Reader follow *that* (or, ideally, just the current rotation, whatever it comes from) rather that each coming up with their own thing.

**Comment by NiLuJe** [comment]
Yup, so, we're back to "let them follow the current rota" and nothing fancier than that ;).

**Comment by NiLuJe** [comment]
> Yup, so, we're back to "let them follow the current rota" and nothing fancier than that ;).

A.K.A., don't do anything, because it's already the right rotation ;p.

**Comment by Frenzie** [comment]
Sounds like the correct behavior to me anyway. ;-)

**Comment by hius07** [comment]
> don't do anything, because it's already the right rotation

That is exactly what I do with Text Editor here but you don't like it.

**Comment by NiLuJe** [comment]
> That is exactly what I do with Text Editor here but you don't like it.

I didn't like the *implementation* bleeding through other widgets, more than the actual behavior ;).

**Comment by hius07** [comment]
Because we have a new feature: manual rotation.

**Comment by hius07** [comment]
That may cause different rotation states of Screen and FM.

**Comment by NiLuJe** [comment]
And it raises a whole lot of conundrums about what to do about it, as we've been discussing ;).

I don't want to it to bleed through other widgets, and I don't want it to be a special snowflake, so I'd just KISS and keep everything in sync (... but implemented in a way that doesn't bleed through other widgets code ;p)

**Comment by hius07** [comment]
> _implementation_ bleeding through other widgets

For my good understanding, can you show me what do you mean.

**Comment by NiLuJe** [comment]
> (But if making it behave slightly nicer doesn't require jumping through too many hoops, go for it).

e.g.,

* On Show, `nil` parent widget's rotation handler
* On CloseWidget, restore parent widget's rotation handler and call it with `Screen:getRotationMode()` to re-sync

**Comment by mergen3107** [comment]
@hius07 
Please let me know when it's ready to test :D

**Comment by hius07** [comment]
It is.
The change is that rotating the device (with gyro) rotates the Text editor but does not rotate underlying FM/Reader.
So after closing TE, FM/Reader are kept in the initial orientation.

**Comment by mergen3107** [comment]
@hius07 
Thanks! Tested, works OK.

However, my observations are following:
1) Rotating with gyro rotates both TE and Reader.
2) Rotating by hand only rotates TE.

**Comment by hius07** [comment]
I tested as follows: the device is 0 degrees, open reader (portrait mode), open TE menu, open any file in TE, rotate the device 90 - TE is rotated. Close TE - the reader is not rotated (still portrait mode) and the menu is still opened.

**Comment by hius07** [comment]
What do you see if you repeat my steps?

**Comment by mergen3107** [comment]
If I do it with gyro, everything is the same, except that Reader with menu is rotate the same as TE.

**Comment by mergen3107** [comment]
I rotate by 180 degrees, because I use that setting to `Lock auto rotation to current orientation`.

**Comment by hius07** [comment]
> `Lock auto rotation to current orientation`.

Can you uncheck this setting and rotate **the device** 90 with opened TE?

**Comment by mergen3107** [comment]
OK, with this setting unchecked, it behaves as you describe: gyro only rotates TE (90 or 180).

**Comment by hius07** [comment]
Thanks! I do not have a gyro device and do not have this setting. I'll check out the codepath to understand how it works.

**Comment by hius07** [comment]
https://github.com/koreader/koreader/blob/f8446538c0d87fc334a8b14640f28093bdb40528/frontend/device/input.lua#L1187-L1200
With this setting activated there is no rotation event, but screen rotation and repainting.
I'm afraid separating the upper widget (Text editor) from all the underlying stuff would be rather tricky. Is it worth? @NiLuJe

**Comment by NiLuJe** [comment]
Yeah, I'll try to take a look tomorrow, can't quite muster the brainpower to grok what kind of interactions y'all are talking about here on a Friday night ;p.

**Comment by hius07** [comment]
Actually, I do not think it's a bothering behavior.
Text editor rotation is useful to see long lines in the landscape mode, and if a user has disabled this orientation, and rotates the device by 180, it means maybe he wants just to change the hold, and it's okay to rotate the Text editor and all the layers under it.
And manual rotation can help in any case.

**Comment by hius07** [comment]
TL;DR I believe it's good now.

**Comment by hius07** [comment]
> `Lock auto rotation to current orientation`.

BTW, @mergen3107 having this setting enabled you cannot use Profiles auto-exec on rotation.

**Comment by mergen3107** [comment]
Thanks! I haven't noticed it yet, because I stayed in Portrait/Portrait for a while. 

Having it unchecked is tough (as I described in another close ticket recently)

I guess I'll revert to my manual complex rotation profiles :D

**Comment by mergen3107** [comment]
> Having it unchecked is tough (as I described in another close ticket recently)

https://github.com/koreader/koreader/issues/12666

**Comment by mergen3107** [comment]
Oh, wait, disregard my comment above :D 

Yes, I won't be able to use auto profiles because I locked 90 deg rotations. 

I left auto rotation profiles set up and now I just use another profiles to just rotate by 90, and it works. 

All good!

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12658)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved

**Comment by hius07** [comment]
The menu rotation buttons in TE work in the same way as in the FM Rotation menu.
And the upper 90 button works in the same way as the leftmost Reader bottom menu button.
Looks like consistent, isn't it?

**Comment by poire-z** [comment]
Kudos for the consistency :)
(I was just mentionning that, for me who have ever only used the bottom menu ones for the last 3 years, and only now tried again these curvy arrows, I was hit by the same ambiguity as I had been before.)
